bindings/rust/bench/src/bin/graph_perf.rs

```diff
@@ -0,0 +1,159 @@
+    tag_names.sort();
+
+    // get the indices of all of the tags for plotting
+    let tag_names_to_index = tag_names
+        .iter()
+        .enumerate()
+        .map(|(i, version)| (version, i))
+        .collect::<HashMap<_, _>>();
+
+    // keep track of ymax for plotting range
+    let mut y_max = 0.0;
+
+    // read in all the data
+    // loop through each record (row in csv), one per tag/version
+    for record in csv_reader.records().map(|res| res.unwrap()) {
+        let mut record_iter = record.into_iter();
+
+        // first item in record is tag
+        let tag = Version::parse(&record_iter.next().unwrap()[1..])?;
+
+        // every two items after is y/y-stderr pair
+        // loop through each header and track (tag_name, y, y_bar)
+        for header in headers.iter() {
+            let (y, y_stderr) = (
+                record_iter.next().unwrap().parse()?,
+                record_iter.next().unwrap().parse::<f64>()?,
+            );
+            if y > y_max {
+                y_max = y;
+            }
+            map.get_mut(header.as_str())
+                .unwrap()
+                .push((tag.clone(), y, y_stderr * 1.96)); // 1.96 is stderr coefficient for 95% confidence interval
+        }
+    }
+
+    // setup plotting
+    let drawing_area =
+        SVGBackend::new("historical-perf/historical-perf.svg", (1000, 500)).into_drawing_area();
+    drawing_area.fill(&WHITE)?;
+
+    let mut ctx = ChartBuilder::on(&drawing_area)
+        .caption(
+            "Handshake performance over versions since Jun 2022",
+            ("sans-serif", 30).into_font(),
+        )
+        .set_label_area_size(LabelAreaPosition::Left, (10).percent()) // axes padding
+        .set_label_area_size(LabelAreaPosition::Bottom, (8).percent())
+        .build_cartesian_2d(
+            (0..num_tags + 1).with_key_points((1..num_tags).step_by(2).collect()), // put labels on every other tag
+            0.0..(1.3 * y_max), // upper y bound on plot is 1.3 * y_max
+        )?;
+
+    ctx.configure_mesh()
+        .light_line_style(RGBAColor(235, 235, 235, 1.0)) // change gridline color
+        .bold_line_style(RGBAColor(225, 225, 225, 1.0))
+        .x_desc("Version") // axes labels
+        .y_desc("Time (ms)")
+        .x_labels(num_tags)
+        .x_label_formatter(&|x| {
+            // change x coord (index of tag in tag_names) to string
+            if let Some(tag_name) = tag_names.get(*x) {
+                tag_name.to_string()
+            } else {
+                "hi".to_string()
+            }
+        })
+        .y_labels(5) // max 5 labels on y axis
+        .y_label_formatter(&|y| format!("{} ms", y / 1000000.0))
+        .draw()?;
+
+    // go through each handshake type and plot them
+    for (i, (bench_name, data)) in map.iter().enumerate() {
+        // remove data that returned error while benching
+        // heuristic: times < 1% of y_max are invalid/had error
+        let filtered_data = data
```

Comment:
Because the invalid data is from handshakes that just didn't happen (the only thing being benched is the `if let Some()` statement), I didn't think it'd be useful to graph that. I just decided to not plot those points that were obviously bad. I know there's a `--no-fail-fast` flag I recently discovered that has `cargo bench` make all the benches run even if one panics, but I don't know if that applies to this case because all of the handshake benches happen in one file; I'll test it out.

Reply:
It seems that once one bench inside the target in this case handshake) panics, the rest don't run, but other bench targets (such as throughput and memory in future features) are finished. Because of this, I think it's still useful to set --no-fail-fast, but keep the heuristic way to tell if benching had an error.

Reply:
So are there specific versions where it always panics and we get no data? Would it make sense to just have specific exceptions for these, rather than continuing on failures? I worry that future versions that we expect to succeed will silently panic and be filtered out without us knowing. Could be misunderstanding this though.

Reply:
There were specific security policies for specific versions that would return an error. By handling the `Result::Err` variant instead of panicking on an error with `Result::unwrap()`, the bench continues to run after the failure. If we didn't continue on failure, after one handshake with one configuration fails, the rest of the handshakes in the bench with different configs wouldn't be run, which isn't ideal. 

In the future, if one of the configs has a new error, the results for that specific config would be obviously missing on the output graph, but other configs that still work wouldn't be affected.